pageserver/src/config.rs

```diff
@@ -106,6 +110,168 @@ pub struct PageServerConf {
     pub remote_storage_config: Option<RemoteStorageConfig>,
 }
 
+// needed to simplify config construction
+struct PageServerConfigBuilder {
+    listen_pg_addr: Option<String>,
+
+    listen_http_addr: Option<String>,
+
+    checkpoint_distance: Option<u64>,
+    checkpoint_period: Option<Duration>,
+
+    gc_horizon: Option<u64>,
+    gc_period: Option<Duration>,
+    superuser: Option<String>,
+
+    page_cache_size: Option<usize>,
+    max_file_descriptors: Option<usize>,
+
+    workdir: Option<PathBuf>,
+
+    pg_distrib_dir: Option<PathBuf>,
+
+    auth_type: Option<AuthType>,
+
+    auth_validation_public_key_path: Option<Option<PathBuf>>,
+    remote_storage_config: Option<Option<RemoteStorageConfig>>,
+
+    id: Option<ZNodeId>,
+}
+
+impl Default for PageServerConfigBuilder {
+    fn default() -> Self {
+        use defaults::*;
+        Self {
+            listen_pg_addr: Some(DEFAULT_PG_LISTEN_ADDR.to_string()),
+            listen_http_addr: Some(DEFAULT_HTTP_LISTEN_ADDR.to_string()),
+            checkpoint_distance: Some(DEFAULT_CHECKPOINT_DISTANCE),
+            checkpoint_period: Some(
+                humantime::parse_duration(DEFAULT_CHECKPOINT_PERIOD)
+                    .expect("cannot parse default checkpoint period"),
+            ),
+            gc_horizon: Some(DEFAULT_GC_HORIZON),
+            gc_period: Some(
+                humantime::parse_duration(DEFAULT_GC_PERIOD)
+                    .expect("cannot parse default gc period"),
+            ),
+            superuser: Some(DEFAULT_SUPERUSER.to_string()),
+            page_cache_size: Some(DEFAULT_PAGE_CACHE_SIZE),
+            max_file_descriptors: Some(DEFAULT_MAX_FILE_DESCRIPTORS),
+            workdir: Default::default(),
+            pg_distrib_dir: Some(PathBuf::new()),
```

Comment:
> `Some(PathBuf::new())`

That is a bit odd to me, is `PathBuf::new()` ever a valid path for the `pg_distrib_dir`?

Reply:
Hmm, I've blindly copied defaults, in the config checking code I see this: https://github.com/zenithdb/zenith/blob/74a0942a773359754e704a662e7d109cd07489a9/pageserver/src/config.rs#L289
so I think I can directly use it as a default without intermediate value